### PR TITLE
Update typescript

### PR DESCRIPTION
**Pull request by mdonnalley**
Updates
- typescript to `4.1.3`
- @typescript-eslint/eslint-plugin to `^4.2.0`
- @typescript-eslint/parser to `^4.2.0`

@W-8667867@

### CODE REVIEW

File: packages/ts-types/src/narrowing/to.ts [hunk 1]
```diff
@@ -39,6 +39,7 @@ export function toAnyJson<T>(value: Nullable<T>, defaultValue: AnyJson): AnyJson
 // underlying function
 export function toAnyJson<T>(value: Nullable<T>, defaultValue?: AnyJson): Optional<AnyJson> {
   try {
+    /* eslint-disable-next-line @typescript-eslint/no-unsafe-return */
```
Comment (shetzel) [inline packages/ts-types/src/narrowing/to.ts:42]: Can't we just coerce the type returned by `JSON.parse()` to be `AnyJson` rather than disable the rule?

**Review by shetzel — CHANGES_REQUESTED** [review]
Lots of great changes in this!  However, I think we can remove more rule disablements.  It should be very rare when we disable rules and there should be a reason provided when we do....unless we're planning to file more stories about removing rule disablements.

Not sure where things are with the lazy loader code.  Maybe @amphro has a better sense of the status of that lib.

**Review by peternhale — APPROVED** [review]

**Review by shetzel — APPROVED** [review]

File: packages/eslint-config-salesforce-typescript/package.json [hunk 2]
```diff
@@ -22,8 +22,8 @@
     "url": "https://github.com/forcedotcom/sfdx-dev-packages/issues"
   },
   "devDependencies": {
-    "@typescript-eslint/eslint-plugin": "^2.30.0",
-    "@typescript-eslint/parser": "^2.30.0",
+    "@typescript-eslint/eslint-plugin": "^4.2.0",
+    "@typescript-eslint/parser": "^4.2.0",
     "eslint": "^6.8.0",
```
Comment (amphro) [inline packages/eslint-config-salesforce-typescript/package.json:27]: Geesh, I wonder if we also need a similar story around keeping eslint up-to-date... it is never ending.

**Review by amphro — APPROVED** [review]
Ya, bummer we need to exclude so many rules. Some of these libraries are so low level that we will probably never completely remove all the exclusions. It might be nice to have comments on ones we will never change and TODOs for the ones we want to fix.

**Comment by amphro** [comment]
I ran and tested this locally, as well as linked it to several projects. The new versions of kit and ts-types works great. The new dev scripts cause a lot of compilation problems in new repos but that is expected with a major bump of typescript, but that is only a dev dep.

**Pull request merged**